#!/usr/bin/env python3

import os
import pickle
import stat
import tempfile
import time
from concurrent import futures

import exploit_pb2
import exploit_pb2_grpc
import grpc
from zeratool_lib import ZeratoolInputStreams, exploit

YEAR_IN_SECONDS = 365 * 60 * 60 * 24


def _create_temp_binary(content: bytes) -> tempfile.NamedTemporaryFile:
    binary = tempfile.NamedTemporaryFile()

    binary.write(content)
    binary.flush()

    os.chmod(binary.name, stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO)

    return binary


class EService(exploit_pb2_grpc.ExploitServiceServicer):
    def Exploit(self, request, _):
        temp_file = _create_temp_binary(request.binary)

        input_stream = ZeratoolInputStreams(request.input_stream)
        overflow_only = request.overflow_only
        format_only = request.format_only
        win_funcs_used = request.serialized_win_funcs.split(",")
        if len(win_funcs_used) == 1 and win_funcs_used[0] == '':
            win_funcs_used = None

        print("temp_file.name: ", temp_file.name)
        print("overflow_only: ", overflow_only)
        print("format_only: ", format_only)
        print("win_funcs_used: ", win_funcs_used)

        result = exploit(
            temp_file.name,
            input_stream,
            overflow_only=overflow_only,
            format_only=format_only,
            win_funcs=win_funcs_used,
        )

        return exploit_pb2.exploit(pickledExploit=pickle.dumps(result))


def serve() -> None:
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=4))
    exploit_pb2_grpc.add_ExploitServiceServicer_to_server(EService(), server)
    server.add_insecure_port("[::]:13000")
    server.start()
    try:
        # server.start() will not block.
        while True:
            time.sleep(YEAR_IN_SECONDS)
    except KeyboardInterrupt:
        server.stop(0)


if __name__ == "__main__":
    serve()
